algorithms.py

```diff
@@ -968,36 +969,42 @@ def get_total_gc_rate(self, label):
 
     def get_total_gc_left_rate(self, label):
         gc_left_total = self.get_total_gc_left(label)
-        # NOTE: this is an approximation. Should we be multiplying by the
-        # local GC rate when iterating over the segments?
-        mean_gc_rate = self.gc_map.mean_rate
-        return mean_gc_rate * self.track_length * gc_left_total
-
-    def get_individual_length(self, head):
-        tail = head
-        while tail.next is not None:
-            tail = tail.next
-        # TODO adjust for discrete/continuous genome here?
-        return tail.right - head.left - 1
+        return gc_left_total
 
     def get_total_gc_left(self, label):
         gc_left_total = 0
-        x = (self.track_length - 1) / self.track_length
-        for pop in self.P:
-            for ind in pop.iter_label(label):
-                dist = self.get_individual_length(ind)
-                gc_left_total += 1 - x ** dist
+        if self.gc_rate_constant:
+            num_ancestors = sum(pop.get_num_ancestors() for pop in self.P)
+            mean_gc_rate = self.gc_map.mean_rate
+            gc_left_total = num_ancestors * mean_gc_rate * self.track_length
+        else:
+            mean_gc_rate = self.gc_map.mean_rate
+            for pop in self.P:
+                for ind in pop.iter_label(label):
+                    # FIXME this is only giving the correct value for gene conversion maps with constant gc rate
+                    # but this is faster calculated above. Keeping this here to adapt it to nonconstant gc rates later:
+                    # We should be using the precomputed local GC left rates when iterating over the individuals.
+                    gc_left_total += mean_gc_rate * self.track_length
         return gc_left_total
 
     def find_cleft_individual(self, label, cleft_value):
         gc_left_total = 0
-        x = (self.track_length - 1) / self.track_length
-        for pop in self.P:
-            for ind in pop.iter_label(label):
-                dist = self.get_individual_length(ind)
-                gc_left_total += 1 - x ** dist
-                if gc_left_total >= cleft_value:
-                    return ind
+        if self.gc_rate_constant:
+            mean_gc_rate = self.gc_map.mean_rate
+            for pop in self.P:
+                num_ancestors = pop.get_num_ancestors()
+                if gc_left_total + num_ancestors * mean_gc_rate * self.track_length >= cleft_value:
```

Comment:
There's a lot of multipling and dividing by constants here. Wouldn't it be simpler to have
```python
individual_index = math.floor(gc_left_value / (mean_gc_rate * self.track_length))
for pop in self.P:
    num_ancestors = pop.get_num_ancestors()
    if individual_index < num_ancestors:
        return pop._ancestors[label][individual_index]
    individual_index -= num_ancestors
```

(I haven't tested this!)